kyaml/setters2/settersutil/substitutioncreator.go

```diff
@@ -62,3 +73,171 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
+			}
+			m[marker] = value
+		} else {
+			// Ex: fv = "samething/ubuntu:0.1.0" pattern = "something/IMAGE:VERSION". Error out at 'a' in fv.
+			if fv[fvInd] != pattern[patternInd] {
+				return nil, errors.Errorf("unable to derive values for markers, create setters for all markers and then try again")
+			}
+			fvInd++
+			patternInd++
+		}
+	}
+	// check if both strings are completely visited or throw error
+	if fvInd < len(fv) || patternInd < len(pattern) {
+		return nil, errors.Errorf("unable to derive values for markers, create setters for all markers and then try again")
+	}
+	return m, nil
+}
+
+// GetStartIndices returns the start indices of all the markers in the pattern
+func (c SubstitutionCreator) GetStartIndices() (map[int]string, error) {
+	indices := make(map[int]string)
+	p := c.Pattern
+	for _, value := range c.Values {
+		m := value.Marker
+		found := false
+		for i := range p {
+			if strings.HasPrefix(p[i:], m) {
+				indices[i] = m
+				found = true
+			}
+		}
+		if !found {
+			return nil, errors.Errorf("unable to find marker " + m + " in the pattern")
+		}
+	}
+	if err := validateMarkerPositions(indices); err != nil {
+		return nil, err
+	}
+	return indices, nil
+}
+
+// validateMarkerPositions takes the indices map and checks if any of 2 markers not have delimiters
+// and returns error
+func validateMarkerPositions(indices map[int]string) error {
+	for k1, v1 := range indices {
+		for k2, v2 := range indices {
+			if k1 != k2 && k1+len(v1) == k2 {
+				return errors.Errorf("markers %s and %s are found to have no delimiters between them", v1, v2)
+			}
+		}
+	}
+	return nil
+}
+
+// extractValueForMarker returns the value string for a marker and the incremented index
+func (c SubstitutionCreator) extractValueForMarker(fvInd int, fv string, patternInd int, indices map[int]string) (string, int) {
+	valueStartInd := fvInd
+	nonMarkerStr := strTillNextMarker(indices, patternInd, c.Pattern)
+
+	// return the remaining string of fv till end if patternInd is at end of pattern
+	if patternInd == len(c.Pattern) {
+		return fv[fvInd:], len(fv)
+	}
+
+	// keep incrementing fvInd till we find the substring between 2 markers in fv
+	// In example fv = "something/ubuntu::0.1.0", pattern = "something/IMAGE::VERSION",
+	// proceed till we find substring "::" in fv which makes value = ubuntu for marker IMAGE
+	for fvInd < len(fv) && substrOfLen(fv, fvInd, len(nonMarkerStr)) != nonMarkerStr {
+		fvInd++
+	}
+	return fv[valueStartInd:fvInd], fvInd
+}
+
+// substrOfLen takes a string, start index and length and returns substring of given length
+// or till end of string
+func substrOfLen(str string, startInd int, length int) string {
+	return str[startInd:min(len(str), startInd+length)]
+}
+
+// lenToNextMarker takes in the indices map, an index and returns the distance to
+// next marker start index in indices map
+func strTillNextMarker(m map[int]string, j int, pattern string) string {
+	res := math.MaxInt32
+	for k := range m {
```

Comment:
// find the index of the next marker (e.g. the smallest value in map greater than j)